profiler/src/main/java/com/splunk/opentelemetry/profiler/allocation/exporter/PprofAllocationEventExporter.java

```diff
@@ -0,0 +1,291 @@
+import java.util.Map;
+import java.util.zip.GZIPOutputStream;
+import jdk.jfr.consumer.RecordedEvent;
+import jdk.jfr.consumer.RecordedFrame;
+import jdk.jfr.consumer.RecordedMethod;
+import jdk.jfr.consumer.RecordedStackTrace;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PprofAllocationEventExporter implements AllocationEventExporter {
+  private static final Logger logger = LoggerFactory.getLogger(PprofAllocationEventExporter.class);
+
+  private final LogProcessor logProcessor;
+  private final Resource resource;
+  private final AllocationDataFormat allocationDataFormat;
+  private ProfileData profileData = new ProfileData();
+
+  private PprofAllocationEventExporter(Builder builder) {
+    this.logProcessor = builder.logProcessor;
+    this.resource = builder.resource;
+    this.allocationDataFormat = builder.allocationDataFormat;
+  }
+
+  @Override
+  public void export(RecordedEvent event, AllocationEventSampler sampler, SpanContext spanContext) {
+    RecordedStackTrace stackTrace = event.getStackTrace();
+    if (stackTrace == null) {
+      return;
+    }
+
+    Profile.Builder profile = profileData.profile;
+    LocationTable locationTable = profileData.locationTable;
+    StringTable stringTable = profileData.stringTable;
+
+    // XXX jdk mission control "TLAB allocations" view seems to use tlabSize as allocation size
+    long allocationSize = event.getLong("allocationSize");
+
+    Sample.Builder sample = Sample.newBuilder();
+    sample.addValue(allocationSize);
+    // XXX StackSerializer limits stack depth. Although I believe jfr also limits it should we
+    // attempt to limit the depth here too?
+    for (RecordedFrame frame : stackTrace.getFrames()) {
+      RecordedMethod method = frame.getMethod();
+      if (method == null) {
+        sample.addLocationId(locationTable.get("", "unknown.unknown", -1));
+      } else {
+        sample.addLocationId(
+            locationTable.get(
+                "", // file name is not known
+                method.getType().getName() + "." + method.getName(),
+                frame.getLineNumber()));
+      }
+    }
+    if (spanContext != null && spanContext.isValid()) {
+      Label.Builder traceIdLabel = Label.newBuilder();
+      traceIdLabel.setKey(stringTable.get("trace_id"));
+      traceIdLabel.setStr(stringTable.get(spanContext.getTraceId()));
+      sample.addLabel(traceIdLabel.build());
+
+      Label.Builder spanIdLabel = Label.newBuilder();
+      spanIdLabel.setKey(stringTable.get("span_id"));
+      spanIdLabel.setStr(stringTable.get(spanContext.getSpanId()));
+      sample.addLabel(spanIdLabel.build());
+
+      Label.Builder eventTypeLabel = Label.newBuilder();
+      eventTypeLabel.setKey(stringTable.get("event_type"));
+      eventTypeLabel.setStr(stringTable.get(event.getEventType().getName()));
+      sample.addLabel(eventTypeLabel.build());
+    }
+    // XXX are there any more attributes that should be added to the sample?
```

Comment:
On the backend, the stack traces are expected to have the following attributes on top of this:

* `service.name` (as set via `otel.service.name`)
* `deployment.environment` (as set via `otel.resource.attributes -> environment`)
* `thread.name`, `thread.id`, `thread.state` (when known)
* `source.event.name` (e.g. "jdk.ObjectAllocationInNewTLAB" or "jdk.ObjectAllocationOutsideTLAB")
* `telemetry.sdk.language`
* `container.id`
* `host.name`

Now, some of these attributes like service name, env, host name, etc, are not specific to a given stack trace, so can be set on the batch level. But we do need to receive them.

@vovencij please review these too.

Reply:
You need to differentiate attributes that are needed once for the whole batch and the ones that are needed for every stack. Here we are dealing with attributes that are needed for every stack.

Reply:
... although thread name, state and ID are per-stack.

Reply:
Yeah, the per-stack attributes are:
* event source/type
* thread name/id/state
* trace/span ids

So looks like only the thread info is missing. The plain text data format does include the name, so let's include that too, even if state and id are unavailable.